### DISCUSSION TITLE
How to use GNPy

### DISCUSSION THREAD

**Discussion by ghost** [Telecominfraproject/oopt-gnpy#378] [Q&A] [answered]
My team is working on an Optical Network Design Project and there aren't enough helping materials available for us to use GNPy. Could any one of the repository maintainers guide us through it?

Our requirements are:

1. Design a physical topology
2. Assign traffic on each link, as well as other parameters
3. Perform Routing and spectrum assignment

**Comment by jktjkt** [comment]
Hi @mohammadsaadraza , thanks for reaching out. You're right that our documentation and tutorials should be improved. Right now the best "first step" is usually trying to play with the [demo files in `gnpy/example-data/`](https://github.com/Telecominfraproject/oopt-gnpy/tree/master/gnpy/example-data).

There are also some training/demo videos that @EstherLerouzic made, these were available via the TIP project's Workplace. Unfortunately, that website has been recently decommissioned. I'll check to see if these videos can be uploaded to a new web site.

In the past months we've been working on a transition to a YANG-defined SW interface which will make it easier to interface with GNPy. In this work, the documentation has been extended with more explanations about GNPy's mode of operation -- what is the "equipment library", how does "the topology" looks like, etc -- essentially, we will have the instructions on how to connect the dots together. This is an ongoing effort, I'll keep you posted.

**Reply by ghost** [reply]
Thank You @jktjkt. I have been playing around with the example code functions and have acquired a high-level understanding of how the library is functioning. But still, I need to figure out how to query path requests / blocked requests at a large scale with GNPy. For that, I need to learn the structure of service.json files, which I can't pick out yet.

If it is possible, the training/demo videos would help a lot. A drive link would be just fine.

As I keep working with GNPy, I'll probably keep this thread alive with questions.

**Comment by jktjkt** [comment, marked as answer]
Here are our old training/demonstration videos:

- [Parts 1+2/5: Using GNPy as a library](https://www.youtube.com/watch?v=plP0Vp0dxLU)
- [Part 3/5](https://www.youtube.com/watch?v=de6nZmPb44I)
- [Part 4/5](https://www.youtube.com/watch?v=F39ikBf4jjU)
- [Part 5/5](https://www.youtube.com/watch?v=yr96_g8zFHM)
- [Demo, June 2018](https://www.youtube.com/watch?v=g1S7k52VkWQ)
- [Demo, Power Sweep, August 2018](https://www.youtube.com/watch?v=IzE6uNmya7E)

Keep in mind that it's been more than two years since these have been done. The command names have changed, and there might be better ways of achieving something with GNPy today. But these are the videos which we have at the moment. Please let us know whether they help.

**Reply by ghost** [reply]
Yes @jktjkt, the videos have been very helpful. Thank you.

I will keep you posted if my team has any further questions.

**Reply by lu0909** [reply]
Thank you for the videos.
But i still have doubt that If I want to calculate a path from "trx Abilene" to "trx Albany", do I need to create a file like simulation. py shown in the first video, or are there already have one in the provided scripts in https://github.com/Telecominfraproject/oopt-gnpy?

**Reply by jktjkt** [reply]
> Thank you for the videos. But i still have doubt that If I want to calculate a path from "trx Abilene" to "trx Albany", do I need to create a file like simulation. py shown in the first video, or are there already have one in the provided scripts in https://github.com/Telecominfraproject/oopt-gnpy?

These older training videos explain all sorts of use cases around GNPy. Since 2018, we've made it easier to just start running simulations right away. You'll have to [install GNPy first](https://gnpy.readthedocs.io/en/master/install.html#installing-the-python-package); once that happens, you should have commands such as `gnpy-transmission-example` and `gnpy-path-request` available.

**Comment by HidekiNishizawa** [comment]
Thank you for your info, these are helpful for us. Examples on GitHub 
site are also helpful.

On 2021/01/15 1:24, Jan Kundrát wrote:
> Here are our old training/demonstration videos:
> 
>   * Parts 1+2/5: Using GNPy as a library
>     <https://www.youtube.com/watch?v=plP0Vp0dxLU>
>   * Part 3/5 <https://www.youtube.com/watch?v=de6nZmPb44I>
>   * Part 4/5 <https://www.youtube.com/watch?v=F39ikBf4jjU>
>   * Part 5/5 <https://www.youtube.com/watch?v=yr96_g8zFHM>
>   * Demo, June 2018 <https://www.youtube.com/watch?v=g1S7k52VkWQ>
>   * Demo, Power Sweep, August 2018
>     <https://www.youtube.com/watch?v=IzE6uNmya7E>
> 
> Keep in mind that it's been more than two years since these have been 
> done. The command names have changed, and there might be better ways of 
> achieving something with GNPy today. But these are the videos which we 
> have at the moment. Please let us know whether they help.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Telecominfraproject/oopt-gnpy/discussions/378#discussioncomment-282301>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AG2QIVUGWJLZ3RFUIGOHTG3SZ4LFLANCNFSM4V4H6ULA>.
> 

-- 
-----------------------------------------
Hideki Nishizawa
NTT Network Innovation Laboratories
Hideki.nishizawa.zw@hco.ntt.co.jp
Office/Mobile: +81-46-859-2469

**Comment by indayara** [comment]
Dear,
I have questions about the Gnpy .

1) When I run the json script referent a network model, the Gnpy shows transmission results as OSNR ASE (0.1nm, dB)  and SNR_total (0.1nm, dB).  Let me know if I can consider the  nonlinearities penalties as being: OSNR ASE (0.1nm, dB)  -  SNR_total (0.1nm, dB) ?

2) Through the command --show channels  we can obtain the SNR_nli for each channel.  What is SNR_NLI? 

3) In the folder GNPY/CORE/science_utils.py, there are 2 models to calculate the OSNR_ASE and SNR_total and SNR_NLI: GN model and GGN model. In my simulations I don't know what model is being used... How can I configure it?

Tks in advance!

**Reply by jktjkt** [reply]
Hi, and sorry for missing your questions. The GitHub discussion UI can be confusing at times, so I would prefer if new questions are posted as new threads, that makes it easier to track what was answered and what is pending.

> When I run the json script referent a network model, the Gnpy shows transmission results as OSNR ASE (0.1nm, dB) and SNR_total (0.1nm, dB).  Let me know if I can consider the  nonlinearities penalties as being: OSNR ASE (0.1nm, dB)  -  SNR_total (0.1nm, dB) ?

Yes, in general the contribution of these nonlinearities (over a certain bandwidth) is added like that.

>  Through the command --show channels  we can obtain the SNR_nli for each channel.  What is SNR_NLI?

That's a ratio of the signal power to just the non-linear impairments. If I read the code correctly, this disregards the ASE, so it is probably useful to tell if your network is "badly optimized" and uses too high of a launch power.

> In the folder GNPY/CORE/science_utils.py, there are 2 models to calculate the OSNR_ASE and SNR_total and SNR_NLI: GN model and GGN model. In my simulations I don't know what model is being used... How can I configure it?

We have some work-in-progress code which will change this, but in the released version and in the current `master` branch, it looks to me that the `ggn_spectrally_separated` is used when you use Raman-aware fibers and pass the `--sim-params` option with the default JSON file.
